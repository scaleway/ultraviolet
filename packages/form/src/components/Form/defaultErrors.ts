import type { RequiredErrors } from '../../types'

export const defaultErrors: RequiredErrors = {
  deps: () => '',
  disabled: () => '',
  max: () => '',
  maxDate: () => '',
  maxLength: () => '',
  min: () => '',
  minLength: () => '',
  onBlur: () => '',
  onChange: () => '',
  pattern: () => '',
  required: () => '',
  setValueAs: () => '',
  shouldUnregister: () => '',
  validate: () => '',
  value: () => '',
  valueAsDate: () => '',
  valueAsNumber: () => '',
}
