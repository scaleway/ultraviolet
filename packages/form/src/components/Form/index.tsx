'use client'

import styled from '@emotion/styled'
import type { ReactNode } from 'react'
import type { FieldValues, UseFormReturn } from 'react-hook-form'
import { FormProvider } from 'react-hook-form'
import { ErrorProvider } from '../../providers'
import type { FormErrors } from '../../types'
import { defaultErrors } from './defaultErrors'

type OnSubmitReturn = string | null | undefined | void

type FormProps<TFieldValues extends FieldValues> = {
  children?: ReactNode
  errors: FormErrors
  name?: string
  onSubmit: (data: TFieldValues) => Promise<OnSubmitReturn> | OnSubmitReturn
  methods: UseFormReturn<TFieldValues>
}

const StyledForm = styled.form`
  display: contents;
`

export const Form = <TFieldValues extends FieldValues>({
  children,
  methods,
  errors,
  onSubmit,
  name,
}: FormProps<TFieldValues>) => {
  const handleSubmit = methods.handleSubmit(async values => {
    const result = await onSubmit(values)

    if (result) {
      methods.setError('root.submit', {
        message: result,
        type: 'custom',
      })
    }
  })

  return (
    <FormProvider {...methods}>
      <ErrorProvider errors={{ ...defaultErrors, ...errors }}>
        <StyledForm
          name={name}
          noValidate
          onSubmit={async e => {
            e.preventDefault()
            e.stopPropagation()
            await handleSubmit(e)
          }}
        >
          {children}
        </StyledForm>
      </ErrorProvider>
    </FormProvider>
  )
}
