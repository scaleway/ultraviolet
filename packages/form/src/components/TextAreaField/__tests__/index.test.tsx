import { renderHook, screen, waitFor } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { mockFormErrors, renderWithForm, renderWithTheme } from '@utils/test'
import { useForm } from 'react-hook-form'
import { describe, expect, test, vi } from 'vitest'
import { Submit } from '../..'
import { Form } from '../../Form'
import { TextAreaField } from '..'

describe('textAreaField', () => {
  test('should render correctly', () => {
    const { asFragment } = renderWithForm(
      <TextAreaField label="Test" name="test" />,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('should submit with onSubmitEnter prop', async () => {
    const onSubmit = vi.fn()
    const { result } = renderHook(() => useForm<{ test: string }>())

    const { asFragment } = renderWithTheme(
      <Form
        errors={mockFormErrors}
        methods={result.current}
        onSubmit={onSubmit}
      >
        <TextAreaField label="Test" name="test" submitOnEnter />
        <Submit>Submit</Submit>
      </Form>,
    )

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalledTimes(0)
    })
    const textareaInput = screen.getByLabelText('Test')
    await userEvent.type(textareaInput, 'This is an example{Enter}')
    await waitFor(() => {
      expect(onSubmit.mock.calls[0][0]).toEqual({
        test: 'This is an example',
      })
    })
    expect(asFragment()).toMatchSnapshot()
  })

  test('should render correctly generated', async () => {
    const onSubmit = vi.fn()
    const { result } = renderHook(() => useForm<{ test: string }>())

    const { asFragment } = renderWithTheme(
      <Form
        errors={mockFormErrors}
        methods={result.current}
        onSubmit={onSubmit}
      >
        <TextAreaField clearable label="Test" name="test" required />
        <Submit>Submit</Submit>
      </Form>,
    )

    await userEvent.click(screen.getByText('Submit'))
    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalledTimes(0)
    })
    const textareaInput = screen.getByLabelText('Test')
    await userEvent.type(textareaInput, 'This is an example')
    await userEvent.click(screen.getByText('Submit'))
    await waitFor(() => {
      expect(onSubmit.mock.calls[0][0]).toEqual({
        test: 'This is an example',
      })
    })
    expect(asFragment()).toMatchSnapshot()
  })
})
