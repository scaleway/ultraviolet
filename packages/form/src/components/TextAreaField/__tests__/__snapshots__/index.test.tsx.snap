// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextAreaField > should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-2 {
  color: #222638;
  font-size: 1rem;
  font-family: Inter,Asap,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.5rem;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-4 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-6 {
  width: 100%;
  resize: vertical;
  background: #ffffff;
  border: 1px solid #d9dadd;
  color: #3f4250;
  border-radius: 0.25rem;
  padding: 0.75rem 0.5rem 0.75rem 1rem;
  padding-right: calc(1.5rem + 0px + 0px);
}

.emotion-6::-webkit-input-placeholder {
  color: #727683;
}

.emotion-6::-moz-placeholder {
  color: #727683;
}

.emotion-6:-ms-input-placeholder {
  color: #727683;
}

.emotion-6::placeholder {
  color: #727683;
}

.emotion-6[data-success='true'] {
  border-color: #22674e;
}

.emotion-6[data-error='true'] {
  border-color: #b3144d;
}

.emotion-6[data-readonly='true'] {
  background: #f9f9fa;
  border-color: #d9dadd;
}

.emotion-6:disabled {
  background: #f3f3f4;
  border-color: #e9eaeb;
  color: #b5b7bd;
}

.emotion-6:disabled::-webkit-input-placeholder {
  color: #b5b7bd;
}

.emotion-6:disabled::-moz-placeholder {
  color: #b5b7bd;
}

.emotion-6:disabled:-ms-input-placeholder {
  color: #b5b7bd;
}

.emotion-6:disabled::placeholder {
  color: #b5b7bd;
}

.emotion-6:not(:disabled):hover {
  border-color: #8c40ef;
}

.emotion-6:not(:disabled):focus {
  outline: none;
  border-color: #8c40ef;
  box-shadow: 0px 0px 0px 3px #8c40ef40;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  position: absolute;
  top: 0.75rem;
  right: 0.5rem;
}

<div
    data-testid="testing"
  >
    <form
      novalidate=""
    >
      <div
        class="emotion-0 emotion-1"
      >
        <label
          class="emotion-2 emotion-3"
          for=":r0:"
        >
          Test
        </label>
        <div
          class="emotion-4 emotion-5"
        >
          <textarea
            aria-invalid="false"
            class="emotion-6 emotion-7"
            data-error="false"
            data-readonly="false"
            data-success="false"
            id=":r0:"
            name="test"
            rows="3"
          />
          <div
            class="emotion-8 emotion-9"
          />
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`TextAreaField > should render correctly generated 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25rem;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-4 {
  color: #222638;
  font-size: 1rem;
  font-family: Inter,Asap,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.5rem;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-6 {
  color: #b3144d;
  font-size: 1rem;
  font-family: Inter,Asap,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.5rem;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-8 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-10 {
  width: 100%;
  resize: vertical;
  background: #ffffff;
  border: 1px solid #d9dadd;
  color: #3f4250;
  border-radius: 0.25rem;
  padding: 0.75rem 0.5rem 0.75rem 1rem;
  padding-right: calc(1.5rem + 300px + 0px);
}

.emotion-10::-webkit-input-placeholder {
  color: #727683;
}

.emotion-10::-moz-placeholder {
  color: #727683;
}

.emotion-10:-ms-input-placeholder {
  color: #727683;
}

.emotion-10::placeholder {
  color: #727683;
}

.emotion-10[data-success='true'] {
  border-color: #22674e;
}

.emotion-10[data-error='true'] {
  border-color: #b3144d;
}

.emotion-10[data-readonly='true'] {
  background: #f9f9fa;
  border-color: #d9dadd;
}

.emotion-10:disabled {
  background: #f3f3f4;
  border-color: #e9eaeb;
  color: #b5b7bd;
}

.emotion-10:disabled::-webkit-input-placeholder {
  color: #b5b7bd;
}

.emotion-10:disabled::-moz-placeholder {
  color: #b5b7bd;
}

.emotion-10:disabled:-ms-input-placeholder {
  color: #b5b7bd;
}

.emotion-10:disabled::placeholder {
  color: #b5b7bd;
}

.emotion-10:not(:disabled):hover {
  border-color: #8c40ef;
}

.emotion-10:not(:disabled):focus {
  outline: none;
  border-color: #8c40ef;
  box-shadow: 0px 0px 0px 3px #8c40ef40;
}

.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  position: absolute;
  top: 0.75rem;
  right: 0.5rem;
}

.emotion-14 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 1.5rem;
  padding: 0 0.25rem;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 0.25rem;
  border-radius: 0.25rem;
  box-sizing: border-box;
  width: 1.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  outline-offset: 2px;
  white-space: nowrap;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 0.875rem;
  font-family: Inter,Asap,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.25rem;
  paragraph-spacing: 0;
  text-case: none;
  background: none;
  border: none;
  color: #3f4250;
}

.emotion-14:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-14:active {
  box-shadow: 0px 0px 0px 3px #151a2d5c;
}

.emotion-14:hover,
.emotion-14:active {
  background: #e9eaeb;
  color: #222638;
}

.emotion-16 {
  vertical-align: middle;
  fill: currentColor;
  height: 16px;
  width: 16px;
  min-width: 16px;
  min-height: 16px;
}

.emotion-16 .fillStroke {
  stroke: currentColor;
  fill: none;
}

.emotion-18 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 3rem;
  padding: 0 1rem;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 0.5rem;
  border-radius: 0.25rem;
  box-sizing: border-box;
  width: auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  outline-offset: 2px;
  white-space: nowrap;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1rem;
  font-family: Inter,Asap,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.5rem;
  paragraph-spacing: 0;
  text-case: none;
  background: #8c40ef;
  border: none;
  color: #ffffff;
}

.emotion-18:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-18:active {
  box-shadow: 0px 0px 0px 3px #8c40ef40;
}

.emotion-18:hover,
.emotion-18:active {
  background: #792dd4;
  color: #f9f9fa;
}

<div
    data-testid="testing"
  >
    <form
      novalidate=""
    >
      <div
        class="emotion-0 emotion-1"
      >
        <div
          class="emotion-2 emotion-1"
        >
          <label
            class="emotion-4 emotion-5"
            for=":rq:"
          >
            Test
          </label>
          <label
            class="emotion-6 emotion-5"
            for=":rq:"
          >
            *
          </label>
        </div>
        <div
          class="emotion-8 emotion-9"
        >
          <textarea
            aria-invalid="false"
            class="emotion-10 emotion-11"
            data-error="false"
            data-readonly="false"
            data-success="false"
            id=":rq:"
            name="test"
            rows="3"
          >
            This is an example
          </textarea>
          <div
            class="emotion-12 emotion-13"
          >
            <button
              aria-label="clear value"
              class="emotion-14 emotion-15"
              type="button"
            >
              <svg
                class="emotion-16 emotion-17"
                viewBox="0 0 16 16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.28 4.22a.75.75 0 0 0-1.06 1.06L6.94 8l-2.72 2.72a.75.75 0 1 0 1.06 1.06L8 9.06l2.72 2.72a.75.75 0 1 0 1.06-1.06L9.06 8l2.72-2.72a.75.75 0 0 0-1.06-1.06L8 6.94z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <button
        class="emotion-18 emotion-19"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`TextAreaField > should submit with onSubmitEnter prop 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-2 {
  color: #222638;
  font-size: 1rem;
  font-family: Inter,Asap,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.5rem;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-4 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-6 {
  width: 100%;
  resize: vertical;
  background: #ffffff;
  border: 1px solid #d9dadd;
  color: #3f4250;
  border-radius: 0.25rem;
  padding: 0.75rem 0.5rem 0.75rem 1rem;
  padding-right: calc(1.5rem + 0px + 0px);
}

.emotion-6::-webkit-input-placeholder {
  color: #727683;
}

.emotion-6::-moz-placeholder {
  color: #727683;
}

.emotion-6:-ms-input-placeholder {
  color: #727683;
}

.emotion-6::placeholder {
  color: #727683;
}

.emotion-6[data-success='true'] {
  border-color: #22674e;
}

.emotion-6[data-error='true'] {
  border-color: #b3144d;
}

.emotion-6[data-readonly='true'] {
  background: #f9f9fa;
  border-color: #d9dadd;
}

.emotion-6:disabled {
  background: #f3f3f4;
  border-color: #e9eaeb;
  color: #b5b7bd;
}

.emotion-6:disabled::-webkit-input-placeholder {
  color: #b5b7bd;
}

.emotion-6:disabled::-moz-placeholder {
  color: #b5b7bd;
}

.emotion-6:disabled:-ms-input-placeholder {
  color: #b5b7bd;
}

.emotion-6:disabled::placeholder {
  color: #b5b7bd;
}

.emotion-6:not(:disabled):hover {
  border-color: #8c40ef;
}

.emotion-6:not(:disabled):focus {
  outline: none;
  border-color: #8c40ef;
  box-shadow: 0px 0px 0px 3px #8c40ef40;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: normal;
  -ms-flex-pack: normal;
  -webkit-justify-content: normal;
  justify-content: normal;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  position: absolute;
  top: 0.75rem;
  right: 0.5rem;
}

.emotion-10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 3rem;
  padding: 0 1rem;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 0.5rem;
  border-radius: 0.25rem;
  box-sizing: border-box;
  width: auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  outline-offset: 2px;
  white-space: nowrap;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1rem;
  font-family: Inter,Asap,sans-serif;
  font-weight: 500;
  letter-spacing: 0;
  line-height: 1.5rem;
  paragraph-spacing: 0;
  text-case: none;
  background: #8c40ef;
  border: none;
  color: #ffffff;
}

.emotion-10:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-10:active {
  box-shadow: 0px 0px 0px 3px #8c40ef40;
}

.emotion-10:hover,
.emotion-10:active {
  background: #792dd4;
  color: #f9f9fa;
}

<div
    data-testid="testing"
  >
    <form
      novalidate=""
    >
      <div
        class="emotion-0 emotion-1"
      >
        <label
          class="emotion-2 emotion-3"
          for=":r3:"
        >
          Test
        </label>
        <div
          class="emotion-4 emotion-5"
        >
          <textarea
            aria-invalid="false"
            class="emotion-6 emotion-7"
            data-error="false"
            data-readonly="false"
            data-success="false"
            id=":r3:"
            name="test"
            rows="3"
          >
            This is an example
          </textarea>
          <div
            class="emotion-8 emotion-9"
          />
        </div>
      </div>
      <button
        class="emotion-10 emotion-11"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
